File: ruby/order/generate_packing_slip_all_d_c.rb
require 'ultracart_api'
require_relative '../constants'

# OrderApi.generatePackingSlipAllDC() is a method that might be used by a fulfillment center or distribution
# center to generate a packing slip to include with a shipment.  This method will return a packing slip for
# an order for all distribution centers involved.
order_api = UltracartClient::OrderApi.new_using_api_key(Constants::API_KEY)

order_id = 'DEMO-0009104390'

begin
  api_response = order_api.generate_packing_slip_all_dc(order_id)

  # Check for errors
  if api_response.error
    puts "Developer Message: #{api_response.error.developer_message}"
    puts "User Message: #{api_response.error.user_message}"
    exit
  end

  # The packing slip will return as a base64 encoded
  # unpack, save off, email, whatever.
  base64_packing_slip = api_response.pdf_base64

  puts base64_packing_slip
rescue StandardError => e
  puts "An error occurred: #{e.message}"
end

File: javascript/order/generatePackingSlipAllDC.js
import { orderApi } from '../api.js';
import fs from 'fs/promises';
import { Buffer } from 'buffer';

export class GeneratePackingSlipAllDC {
    /**
     * OrderApi.generatePackingSlipAllDC() is a method that might be used by a fulfillment center or distribution
     * center to generate a packing slip to include with a shipment.  This method will return a packing slip for
     * an order for all distribution centers involved.
     */
    static async execute() {
        const orderId = 'DEMO-0009104390';

        try {
            // Generate packing slip for all distribution centers
            const apiResponse = await new Promise((resolve, reject) => {
                orderApi.generatePackingSlipAllDC(
                    orderId
                , function (error, data, response) {
                    if (error) {
                        reject(error);
                    } else {
                        resolve(data);
                    }
                });
            });

            // Check for errors
            if (apiResponse.error) {
                console.error('Developer Message:', apiResponse.error.developer_message);
                console.error('User Message:', apiResponse.error.user_message);
                throw new Error('Failed to generate packing slip');
            }

            // the packing slip will return as a base64 encoded
            // unpack, save off, email, whatever.
            const base64PackingSlipOrUndefined = apiResponse.pdfBase64;
            if (base64PackingSlipOrUndefined !== undefined) {
                const base64PackingSlip = base64PackingSlipOrUndefined;
                console.log(base64PackingSlip);

                // Decode Base64 string into a buffer
                const pdfBuffer = Buffer.from(base64PackingSlip, 'base64');

                // Save the buffer to a PDF file
                await fs.writeFile('packing_slip.pdf', pdfBuffer);

                console.log("PDF file saved successfully as 'packing_slip.pdf'");
            }
        } catch (error) {
            console.error('Error generating packing slip:', error);
            process.exit(1);
        }
    }
}

// Optional: If you want to call the method
// GeneratePackingSlipAllDC.execute();